optuna/trial/_frozen.py

```diff
@@ -2,16 +2,28 @@
 from typing import Any
 from typing import Dict
 from typing import Optional
+from typing import Sequence
+from typing import Union
 
 from optuna._experimental import experimental
+from optuna import distributions
 from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.distributions import DiscreteUniformDistribution
+from optuna.distributions import IntLogUniformDistribution
+from optuna.distributions import IntUniformDistribution
+from optuna.distributions import LogUniformDistribution
+from optuna.distributions import UniformDistribution
 from optuna import logging
+from optuna.trial._base import BaseTrial
 from optuna.trial._state import TrialState
 
 _logger = logging.get_logger(__name__)
 
+CategoricalChoiceType = Union[None, bool, int, float, str]
 
-class FrozenTrial(object):
+
+class FrozenTrial(BaseTrial):
```

Comment:
Should we update the `FrozenTrial` in the `structs` submodule as well? It's under deprecation now but not removed until v3.0. I'm thinking that it'd be easier to maintain the code if they're identical. Maybe we can DRY the implementation with something along the following?

```python
# optuna/structs.py
FrozenTrial = deprecated(
    "1.4.0",
    text=(
        "This class was moved to :mod:`~optuna.trial`. Please use "
        ":class:`~optuna.trial.FrozenTrial` instead."
    ),
)(optuna.trial.FrozenTrial)
```

Reply:
I certainly agree with you. Your suggested code looks great!

Reply:
Oh, that was a quick fix, thanks. Just FYI, I didn't verify it locally so not 100% sure that it will work as expected (e.g. that it doesn't alter the original docstrings).

Reply:
At least `test_structs.py` works fine, but another test (`test_study.test_optimize_parallel_storage_warning`) starts to fail... I will investigate it.

Reply:
It might be altering the original, having side-effects on other tests having `optuna.trial.FrozenTrial.__init__` raising warnings as well. Maybe this will do?

```python
@deprecated(
    "1.4.0",
    text=(
        "This class was moved to :mod:`~optuna.trial`. Please use "
        ":class:`~optuna.trial.FrozenTrial` instead."
    ),
)
class FrozenTrial(optuna.trial.FrozenTrial):
    pass
```

Reply:
Your suggestion works! I really thank you for your help!